Internal compiler error in mathfuncs_core.cpp(1239) when enabling AVX

##### System information (version)
- OpenCV => Newest GitHub master from today
- Operating System / Platform =>Windows 8.1 64 Bit
- Compiler => Visual Studio 2015 Update 3

##### Detailed description
I get an internal compiler error (ICE) when building opencv_core in Release mode, on above system.

> 1>c:\opencv\opencv-git\opencv\modules\core\src\mathfuncs_core.cpp(1239): fatal error C1001: An internal error has occurred in the compiler.

The error appears as soon as you enable `ENABLE_AVX`. Disabling AVX and it works again. Debug build works fine as well.
My CPU supports AVX (i7-4700MQ).

##### Steps to reproduce

Compile opencv_core from GitHub master with `ENABLE_AVX` on VS2015 update 3, 64-bit, Release build.

Disabling IPP didn't help. I tried disabling pretty much everything else too (OpenMP, OpenCL, etc.) and it didn't help.

--- a/modules/core/src/mathfuncs_core.cpp
+++ b/modules/core/src/mathfuncs_core.cpp
@@ -373,6 +373,62 @@ void sqrt64f(const double* src, double* dst, int len)
         dst[i] = std::sqrt(src[i]);
 }
 
+// Workaround for ICE in MSVS 2015 update 3 (issue #7795)
+// CV_AVX is not used here, because generated code is faster in non-AVX mode.
+// (tested with disabled IPP on i5-6300U)
+#if (defined _MSC_VER && _MSC_VER >= 1900)
+void exp32f(const float *src, float *dst, int n)
+{
+    CV_INSTRUMENT_REGION()
+
+    CALL_HAL(exp32f, cv_hal_exp32f, src, dst, n);
+    CV_IPP_RUN_FAST(CV_INSTRUMENT_FUN_IPP(ippsExp_32f_A21, src, dst, n) >= 0);
+
+    for (int i = 0; i < n; i++)
+    {
+        dst[i] = std::exp(src[i]);
+    }
+}
+
+void exp64f(const double *src, double *dst, int n)
+{
+    CV_INSTRUMENT_REGION()
+
+    CALL_HAL(exp64f, cv_hal_exp64f, src, dst, n);
+    CV_IPP_RUN_FAST(CV_INSTRUMENT_FUN_IPP(ippsExp_64f_A50, src, dst, n) >= 0);
+
+    for (int i = 0; i < n; i++)
+    {
+        dst[i] = std::exp(src[i]);
+    }
+}
+
+void log32f(const float *src, float *dst, int n)
+{
+    CV_INSTRUMENT_REGION()
+
+    CALL_HAL(log32f, cv_hal_log32f, src, dst, n);
+    CV_IPP_RUN_FAST(CV_INSTRUMENT_FUN_IPP(ippsLn_32f_A21, src, dst, n) >= 0);
+
+    for (int i = 0; i < n; i++)
+    {
+        dst[i] = std::log(src[i]);
+    }
+}
+void log64f(const double *src, double *dst, int n)
+{
+    CV_INSTRUMENT_REGION()
+
+    CALL_HAL(log64f, cv_hal_log64f, src, dst, n);
+    CV_IPP_RUN_FAST(CV_INSTRUMENT_FUN_IPP(ippsLn_64f_A50, src, dst, n) >= 0);
+
+    for (int i = 0; i < n; i++)
+    {
+        dst[i] = std::log(src[i]);
+    }
+}
+#else
+
 ////////////////////////////////////// EXP /////////////////////////////////////
 
 typedef union
@@ -1465,6 +1521,8 @@ void log64f( const double *x, double *y, int n )
         y[i] = y0;
     }
 }
+
+#endif // issue 7795
 
 //=============================================================================
 // for compatibility with 3.0

